trie/database.go

```diff
@@ -289,12 +362,69 @@ func NewDatabaseWithCache(diskdb ethdb.Database, cache int) *Database {
 			HardMaxCacheSize:   cache,
 		})
 	}
-	return &Database{
+	db := &Database{
 		diskdb:    diskdb,
+		noprune:   make(map[common.Hash]struct{}),
 		cleans:    cleans,
-		dirties:   map[common.Hash]*cachedNode{{}: {}},
+		dirties:   map[string]*cachedNode{metaRoot: {}},
 		preimages: make(map[common.Hash][]byte),
 	}
+	if prune {
+		db.pruner = newPruner(db)
+	}
+	return db
+}
+
+// ResumePruning permits the pruner to continue deleting unreferenced trie nodes.
+// It is essential to only ever resume pruning after all data is commited, capped
+// and properly referenced, otherwise the pruner might delete data that's *going-
+// to-be* referenced.
+func (db *Database) ResumePruning() {
+	if db.pruner != nil {
+		atomic.StoreUint32(&db.pruning, 1)
+		db.pruner.resume()
+	}
+}
+
+// PausePruning waits until the pruner is done with processing its current task
+// and then pauses it so a new trie might be properly integrated into the dirty
+// caches and reference counts.
+func (db *Database) PausePruning() {
+	if db.pruner != nil {
+		atomic.StoreUint32(&db.pruning, 0)
```

Comment:
That looks racy. You should probably not clear the `db.pruning` until after `db.pruner.pause()` returns...?

Reply:
Actually, it's racy even so, since the `db.pruner.pause()` only sets another atomic, doesn't actually guarantee that pruning has been paused by the time it returns.

Reply:
That's actually fine, because everything that requires pruning to be stopped starts by obtaining a lock that the pruner is holding. You won't be able to get the lock before it actually pauses.

Calling pause also just sets an interrupt on the pruner, it doesn't block until it returns because the pruner will sill flush to disk after it unlocks the mutex.

Also as far as I see it, I only used this particular atomic for sanity panics.